go/private/rules/transition.bzl

```
@@ -173,6 +195,23 @@ def _go_transition_impl(settings, attr):
         linkmode_label = filter_transition_label("@io_bazel_rules_go//go/config:linkmode")
         settings[linkmode_label] = linkmode
 
+    for setting, value in settings.items():
+        original_value_setting = _original_value_setting(setting)
+        if not original_value_setting:
+            continue
+
+        # If the outgoing configuration would differ from the incoming one in a
+        # value, record the old value in the special original_* setting so that
+        # the real setting can be reset to this value before the new
+        # configuration would cross a non-deps dependency edge.
+        if value != original_settings[setting]:
+            # Encoding as JSON makes it possible to embed settings of arbitrary
+            # types (currently bool, string and string_list) into a string
+            # setting.
+            settings[original_value_setting] = json.encode(original_settings[setting])
```

Why can we not using the original type like:
```suggestion
            settings[original_value_setting] = original_settings[setting]
```